topic-operator/src/main/java/io/strimzi/operator/topic/TopicSerialization.java

```diff
@@ -45,52 +48,47 @@
+            for (String error : errorsInConfig) {
+                LOGGER.warnCr(reconciliation, "KafkaTopic {}/{} has invalid spec.config: {}",
+                        kafkaTopic.getMetadata().getNamespace(),
+                        kafkaTopic.getMetadata().getName(),
+                        error);
+            }
+
+            throw new InvalidTopicException(kafkaTopic, "KafkaTopic " +
+                    kafkaTopic.getMetadata().getNamespace() + "/" + kafkaTopic.getMetadata().getName() +
+                    " has invalid spec.config: " +
+                    String.join(", ", errorsInConfig));
+        }
+    }
+
     @SuppressWarnings("unchecked")
-    private static Map<String, String> topicConfigFromTopicConfig(KafkaTopic kafkaTopic) {
+    private static Map<String, String> topicConfigFromTopicConfig(KafkaTopic kafkaTopic, Reconciliation reconciliation, String kafkaVersion) {
         if (kafkaTopic.getSpec().getConfig() != null) {
-            Map<String, String> result = new HashMap<>(kafkaTopic.getSpec().getConfig().size());
-            for (Map.Entry<String, Object> entry : kafkaTopic.getSpec().getConfig().entrySet()) {
-                String key = entry.getKey();
-                Object v = entry.getValue();
-                boolean isNumberType = v instanceof Long
-                        || v instanceof Integer
-                        || v instanceof Short
-                        || v instanceof Double
-                        || v instanceof Float;
-                if (v instanceof String
-                        || isNumberType
-                        || v instanceof Boolean) {
-                    result.put(key, v.toString());
-                } else {
-                    String msg = "The value corresponding to the key must have a string, number or boolean value";
-                    if (v == null) {
-                        msg += " but the value was null";
-                    } else {
-                        msg += " but was of type " + v.getClass().getName();
-                    }
-                    throw new InvalidTopicException(kafkaTopic, "KafkaTopic's spec.config has invalid entry: " +
-                            "The key '" + key + "' of the topic config is invalid: " + msg);
-                }
-            }
-            return result;
+            KafkaTopicConfiguration cfg = new KafkaTopicConfiguration(reconciliation, kafkaTopic.getSpec().getConfig().entrySet());
+            validateConfiguration(reconciliation, kafkaTopic, kafkaVersion, cfg);
+            return cfg.asOrderedProperties().asMap();
         } else {
             return Collections.EMPTY_MAP;
         }
     }
 
-
     /**
      * Create a Topic to reflect the given KafkaTopic resource.
      * @throws InvalidTopicException
      */
-    public static Topic fromTopicResource(KafkaTopic kafkaTopic) {
+    public static Topic fromTopicResource(KafkaTopic kafkaTopic, Reconciliation reconciliation, String kafkaVersion) {
```

Comment:
I guess that, although it's obtainable from the environment, you're using an explicit parameter for testing purposes? It would make the patch smaller to just overload `fromTopicResource(KafkaTopic)` with `fromTopicResource(KafkaTopic, String)` (I'm assuming that the logging and `KafkaTopicConfiguration` comments mean that you don't need to pass the Reconciliation either).